espnet2/asr/encoder/contextual_block_conformer_encoder.py

```diff
@@ -208,17 +208,19 @@ def forward(
         ilens: torch.Tensor,
         prev_states: torch.Tensor = None,
         is_final=True,
+        infer=False,
     ) -> Tuple[torch.Tensor, torch.Tensor, Optional[torch.Tensor]]:
         """Embed positions in tensor.
 
         Args:
             xs_pad: input tensor (B, L, D)
             ilens: input length (B)
             prev_states: Not to be used now.
+            infer: whether to be used for inference.
         Returns:
             position embedded tensor and mask
         """
-        if self.training or xs_pad.size(0) > 1:
+        if self.training or not infer:

```

Comment:
@D-Keqi 
You introduced here an additional parameter that indicates training mode.
Usually, the pythonic way of signalling that the module is used for evaluation and not for training is by calling [`module.eval()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.eval). This then sets `self.training` to False.
I I read correctly, `eval()` was already called in `espnet2/bin/asr_inference_streaming.py`.

Reply:
Thanks! During training and validation, the train_forward should be used; and during decoding, the forward_infer should be used. The additional parameter "infer" is actually employed to distinguish between validation and inference.

Reply:
Are there other solutions or can we add more precise comments?
This logic is complicated, and it would have a risk of introducing further errors.

Also, can you make a unit/integration test of this mode?
Otherwise, this part would not be stable.

Reply:
> Are there other solutions or can we add more precise comments? This logic is complicated, and it would have a risk of introducing further errors.
> 
> Also, can you make a unit/integration test of this mode? Otherwise, this part would not be stable.

Before making a PR, I always make a test to make sure the code works (at least for me). But it seems that previous implementations do not work for everyone. Inspired by https://github.com/espnet/espnet/issues/3803, I have fixed another issue and I think this version will work. And I have added more comments for clarity.

Reply:
Can you make a test to pass this mode?
You can find an example in 
https://github.com/espnet/espnet/tree/master/test/espnet2
or
https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh

Your local test is important, but this does not cover everyone's usage.

Reply:
> Can you make a test to pass this mode? You can find an example in https://github.com/espnet/espnet/tree/master/test/espnet2 or https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh
> 
> Your local test is important, but this does not cover everyone's usage.

Just let you know that this mode has passed the integration test, and according to other people's responses https://github.com/espnet/espnet/issues/3803#issuecomment-1000020642, I think this version is ok.